### PR TITLE
Remove deprecated Block APIs

### PR DESCRIPTION
**Pull request by dnhatn**
This PR removes deprecated block APIs that use on-breaking block factories. It also changes the Block implementation to the package level to ensure that callers create Blocks via BlockFactory.

More follow-ups are coming.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    /**
+     * A {@link BigArrays} that won't throw {@link CircuitBreakingException}.
+     * <p>
+     * Rather than using the {@link NoneCircuitBreakerService} we use a
+     * very large limit so tests can call {@link CircuitBreaker#getUsed()}.
+     * </p>
+     */
+    protected final BigArrays nonBreakingBigArrays() {
+        return new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, ByteSizeValue.ofBytes(Integer.MAX_VALUE)).withCircuitBreaking();
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java:44]: I wanted to question why we don't use `Long.MAX_VALUE` to be truly non-breaking, but then I realized this code was merely moved.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public void allBreakersEmpty() throws Exception {
+        // first check that all big arrays are released, which can affect breakers
+        MockBigArrays.ensureAllArraysAreReleased();
+
+        for (CircuitBreaker breaker : breakers) {
+            for (var factory : blockFactories) {
+                if (factory instanceof MockBlockFactory mockBlockFactory) {
+                    mockBlockFactory.ensureAllBlocksAreReleased();
+                }
+            }
```
Comment (alex-spies) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java:76]: This code was also only moved, but do you see a reason for the nested for loops? Seems like we check the block factories over and over again, even though check is completely independent of the breaker in the outside loop.

**Review by alex-spies — APPROVED** [review]
LGTM, thanks Nhat!

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java:44]: I think we should not use more than 1GB in tests unless we explicitly want that.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java (hunk 2 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/ComputeTestCase.java]: Yes, I pushed https://github.com/elastic/elasticsearch/pull/103592/commits/d7c4498da1563fd5b63c9f3d37b4fc5db725f3f9.

**Comment by dnhatn** [comment]
Thanks, Alex!

**Pull request merged**